internal/primitives/state-machine/overlayed_entry.go

```diff
@@ -0,0 +1,341 @@
+	}
+}
+
+// Writes a new version of a value.
+// This makes sure that the old version is not overwritten and can be properly
+// rolled back when required.
+func (oe *OverlayedEntry[V]) Set(value StorageValue, firstWriteInTx bool, atExtrinsic *uint32) {
+	var action StorageEntry
+
+	if value == nil {
+		action = RemoveStorageEntry{}
+	} else {
+		action = SetStorageEntry{value}
+	}
+
+	if firstWriteInTx || len(oe.transactions) == 0 {
+		oe.transactions = append(oe.transactions, Transaction[V]{
+			value:      action.(V), //TODO: check this
+			extrinsics: Extrinsics{},
+		})
+	} else {
+		oldValue := oe.ValueRef()
+
+		var setPrev *struct {
+			data                []byte
+			currentMaterialized *uint
+			parentSize          uint
+		}
+
+		switch oldVal := any(oldValue).(type) {
+		case AppendStorageEntry:
+			*oldValue = any(oldVal).(V)
+
+			setPrev = &struct {
+				data                []byte
+				currentMaterialized *uint
+				parentSize          uint
+			}{
+				data:                []byte{},
+				currentMaterialized: oldVal.materializedLength,
+				parentSize:          *oldVal.parentSize,
+			}
+		}
+
+		*oldValue = action.(V) //TODO: check this
+		if setPrev != nil {
+			transactions := len(oe.transactions)
+			if transactions < 2 {
+				panic("`set_prev` is not nil if the value came from parent; qed")
+			}
+
+			parent := &oe.transactions[transactions-2]
+
+			parentValue, ok := any(parent.value).(StorageEntry)
+			if !ok {
+				panic("Parent value is not a storage entry")
+			}
+
+			restoreAppendToParent(parentValue, setPrev.data, setPrev.currentMaterialized, setPrev.parentSize)
+		}
+	}
+
+	if atExtrinsic != nil {
+		oe.TransactionExtrinsics().Insert(*atExtrinsic)
+	}
+}
+
+// Append content to a value, updating a prefixed compact encoded length.
+// This makes sure that the old version is not overwritten and can be properly
+// rolled back when required.
+// This avoid copying value from previous transaction.
+func (oe *OverlayedEntry[V]) Append(
+	element StorageValue,
+	firstWriteInTx bool,
+	init func() StorageValue,
+	atExtrinsic *uint32,
+) {
+	var data []byte
+	var currentLength uint
+	var materializedLength *uint
+	var parentSize *uint
+
+	replace := true
+
+	if len(oe.transactions) == 0 {
+		initValue := init()
+		storageAppend := NewStorageAppend(&initValue)
+
+		// Either the init value is a SCALE list like value to that the `element` gets appended
+		// or the value is reset to `element`.
+		length := storageAppend.ExtractLength()
+		if length != nil {
+			storageAppend.AppendRaw(element)
+			data = initValue
+			currentLength = *length + 1
+			materializedLength = length
+		} else {
+			data = element
+			currentLength = 1
+			materializedLength = nil
+		}
+
+		oe.transactions = append(oe.transactions, Transaction[V]{
+			value: any(&AppendStorageEntry{
+				data:               data,
+				currentLength:      currentLength,
+				materializedLength: materializedLength,
+				parentSize:         nil,
+			}).(V),
```

Comment:
I think if you make `Transaction.Value` attribute just type `StorageEntry` you can remove the type assertion.